### ISSUE TITLE
Ensure that execution controllers are started up before starting workers

### ISSUE THREAD

**Issue by godber** [terascope/teraslice#1649]
We have been seeing cases where Teraslice workers appear to take two minutes to start processing records while other workers have already started doing their work.  This is apparent if you can look at the number of records processed per worker, over time.

If you enable DEBUG logging on the cluster and look at worker logs, you will see the following in the workers that don't start processing immediately:

```
Error: xhr poll error when connecting to ExecutionController at ...
```

We think the cause of this is that some of the workers start up too quickly, before the execution controller is up and accessible ... thus causing their initial requests to timeout.  Workers that come up slightly later, manage to contact the execution controller and begin work right away.

You can see the potential for this here:

https://github.com/terascope/teraslice/blob/b33b821cb9ab6c575972a02699d301b933e1d95b/packages/teraslice/lib/cluster/services/execution.js#L321-L329

In the kubernetes case, the call to `allocateSlicer()` ([ref](https://github.com/terascope/teraslice/blob/master/packages/teraslice/lib/cluster/services/cluster/backends/kubernetes/index.js#L85-L111)) will return as soon as the request to the kubernetes API is completed.  That does not imply that those pods have started and are running however.

**Comment by godber** [comment]
@peterdemartini suggests that this is probably also true in the native clustering case.

**Comment by godber** [comment]
A kubernetes only solution would be to use an init container that blocks startup of the workers until the execution controller is responding:

https://kubernetes.io/docs/concepts/workloads/pods/init-containers/

**Comment by godber** [comment]
There are at least two classes of solutions for this:

* master based polling or event driven solution (implemented in PR above)
* worker init container based solution

While testing failure scenarios I realized one of the disadvantages of the master based solutions is that they will be inherently slower than the worker init container solution.  This is because the worker deployments are not created until after the controller is up, which on minikube has proven to be at least 6s.  After this is up, THEN the worker deployment gets created and its containers will get pulled.  Whereas the init container based solution will result in all of the containers getting pulled at about the same time, and just the starting of worker containers will be blocked until they can reach the execution controller.

The other problem with the master polling the execution controller is that it doesn't test for the ACTUAL problem, which is whether or not the worker can reach the execution controller.  This is still a pretty unlikely/edge case so it's not that big of a deal.  I just keep feeling like it's going to bite us in some unforeseen way.

**Comment by godber** [comment]
In my testing I noticed the following log entry:

```
[2020-02-21T00:32:59.967Z]  INFO: teraslice/13 on teraslice-master-5c7495c8b-k7k4s: execution c1e8f006-2ab4-477b-ab22-b7829a8855e2 is connected (assignment=cluster_master, module=kubernetes_cluster_service, worker_id=BAGSsBpT)
[2020-02-21T00:33:00.574Z]  INFO: teraslice/13 on teraslice-master-5c7495c8b-k7k4s: waitForTcpPortOpen Connected to ts-exc-example-read-stdout-a9d161a6-34ce:45680 (assignment=cluster_master, module=execution_service, worker_id=BAGSsBpT)
```

It is probably possible to use this event to trigger the worker deployment instead of polling.

**Comment by peterdemartini** [comment]
We can use that event but it would mean more of a dependency on socket connections and I don’t think we’d be in a better position. What if instead we move the `waitForTcpPortOpen` check from the cluster master to the messenger client connect method, https://github.com/terascope/teraslice/blob/5b3ce2c0a98ff7390d9b7e1e372817e87a4dba09/packages/teraslice-messaging/src/messenger/client.ts#L111-L114 so that way it will be from the workers perspective and the worker containers will be good to as soon it is available?

**Comment by godber** [comment]
Hopefully this is resolved in https://github.com/terascope/teraslice/pull/1688

**Comment by godber** [comment]
The MR intended to resolve this did not.  I should probably remove the changes put in place to test this once I am certain of my new fix.  Further investigation to the cause of the occasional 2min worker start up delays have revealed the cause to be the execution controller service (proxy) timing out when the pod is not up yet behind that service.  I am now considering removing the service as suggested in https://github.com/terascope/teraslice/issues/1009

Leaving this issue open as a reminder to revert the changes in #1688

**Comment by godber** [comment]
This is fixed as of `v0.64.3`, but that release still has issues, see https://github.com/terascope/teraslice/issues/1831
